app/lib/fields.py

```diff
@@ -305,3 +308,180 @@ def update_choices(
+                self.was_expanded = True  # Always expanded for year-only
+            return result
+        
+        # Handle year-month input (no day)
+        if self.month and not self.day:
+            result = self._fill_year_month_only(year)
+            if result:
+                self.was_expanded = True  # Always expanded for year-month
+            return result
+
+        # Handle full date input
+        if self.day:
+            return self._fill_full_date(year)
+
+        raise CustomValidationError("Please enter a valid date")
+
+    def _validate_year(self):
+        """Validate and return the year as an integer"""
+        if not self.year:
+            raise CustomValidationError(
+                "Year is required if any date component is provided"
+            )
+
+        try:
+            year = int(self.year)
+        except ValueError:
+            raise CustomValidationError("Please enter a valid 4-digit year")
+
+        if not (1000 <= year <= 9999):
+            raise CustomValidationError("Please enter a valid 4-digit year")
+
+        return year
+
+    def _validate_month(self, month):
+        """Validate month is between 1 and 12"""
+        if not (1 <= month <= 12):
+            raise CustomValidationError("Month must be between 1 and 12")
+
+    def _validate_day(self, day):
+        """Validate day is between 1 and 31"""
+        if not (1 <= day <= 31):
+            raise CustomValidationError("Day must be between 1 and 31")
+
+    def _fill_year_only(self, year):
+        """Fill in month and day for year-only dates"""
+        if self.is_from_date:
+            return date(year, 1, 1)  # January 1st
+        else:
+            return date(year, 12, 31)  # December 31st
+
+    def _fill_year_month_only(self, year):
+        """Fill in day for year-month dates"""
+        try:
+            month = int(self.month)
+        except ValueError:
+            raise CustomValidationError(
+                "Month must be a number between 1 and 12"
+            )
+
+        self._validate_month(month)
+
+        if self.is_from_date:
+            return date(year, month, 1)  # First day of month
+        else:
+            last_day = monthrange(year, month)[1]
+            return date(year, month, last_day)  # Last day of month
+
+    def _fill_full_date(self, year):
+        """Create date from full year-month-day input"""
+        if not self.month:
+            raise CustomValidationError("Month is required if day is provided")
+
+        try:
+            day = int(self.day)
+            month = int(self.month)
+        except ValueError:
+            raise CustomValidationError("Day and month must be numbers")
+
+        self._validate_month(month)
+        self._validate_day(day)
+
+        try:
+            return date(year, month, day)
+        except ValueError:
+            raise CustomValidationError(
+                "Invalid date - please check the day is valid for the given month"
+            )
```

Comment:
Do we really need so much validation? 
I think we should start with something simple for the field only ex `Entered date must be a real date, for example Year 2017, Month 9, Day 23` in line with some things done with old etna.